File: src/os/osx/sys-codecs.m
/***********************************************************************
**
**  REBOL [R3] Language Interpreter and Run-time Environment
**
**	Copyright 2021 Oldes
**
**  Licensed under the Apache License, Version 2.0 (the "License");
**  you may not use this file except in compliance with the License.
**  You may obtain a copy of the License at
**
**  http://www.apache.org/licenses/LICENSE-2.0
**
**  Unless required by applicable law or agreed to in writing, software
**  distributed under the License is distributed on an "AS IS" BASIS,
**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
**  See the License for the specific language governing permissions and
**  limitations under the License.
**
************************************************************************
**
**  Title: osx native codecs functionality
**  Author: Oldes
**  Purpose: To load and save images using system API
**	Related: core/n-image.c
**  Note: So far saves just one frame. For multi-image formats it would be
**        better to write a device (image:// port or something like that)
**
************************************************************************/
#include "sys-codecs.h"
#import <CoreGraphics/CGImage.h>
#import <CoreGraphics/CGContext.h>
#import <CoreGraphics/CGDataProvider.h>
#import <ImageIO/ImageIO.h>
#define kSDUTTypeHEIC ((__bridge CFStringRef)@"public.heic")
#define kSDUTTypeHEIF ((__bridge CFStringRef)@"public.heif")
#define kSDUTTypeDDS  ((__bridge CFStringRef)@"com.microsoft.dds")


#ifdef unused
void listImageCodecs(void){
	//puts("Suppported codecs:");
	CFArrayRef mySourceTypes = CGImageSourceCopyTypeIdentifiers();
	CFShow(mySourceTypes);
	CFArrayRef myDestinationTypes = CGImageDestinationCopyTypeIdentifiers();
	CFShow(myDestinationTypes);
}
#endif

int DecodeImageFromFile(const char *uri, unsigned int frame, REBCDI *codi)
{
	int error = 0;
	CFURLRef url = NULL;
	CGImageRef img = NULL;
	CGImageSourceRef imgSrc = NULL;
	UInt32 *pixels;
	CGColorSpaceRef space;
	CGContextRef ctx;
	CFDataRef binSrc;
	
	NSUInteger w, h;
	NSUInteger bytesPerPixel = 4;
	NSUInteger bitsPerComponent = 8;
	
	do {
		if(uri) {
			// decoding from file
			url = urlFromCString(uri);
			imgSrc = CGImageSourceCreateWithURL(url, 0);
			ASSERT_NOT_NULL(imgSrc, 2, "read source file");
			// so far we are decoding only the first image of the source
		} else {
			// decoding from raw data
			binSrc = CFDataCreateWithBytesNoCopy(NULL, codi->data, codi->len, NULL);
			imgSrc = CGImageSourceCreateWithData(binSrc, 0);
		}
		img = CGImageSourceCreateImageAtIndex(imgSrc, 0, 0);
		ASSERT_NOT_NULL(img, 3, "create an image");
		w = CGImageGetWidth(img);
		h = CGImageGetHeight(img);
		pixels = (UInt32*)malloc(w * h * 4); // Rebol's library side must free it!
		ASSERT_NOT_NULL(pixels, 4, "allocate pixels buffer");
		space = CGColorSpaceCreateDeviceRGB();
		ctx = CGBitmapContextCreate(pixels, w, h, bitsPerComponent, bytesPerPixel * w, space, kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
		ASSERT_NOT_NULL(ctx, 5, "create a bitmap context");
		CGContextDrawImage(ctx, CGRectMake(0, 0, w, h), img);
		CGColorSpaceRelease(space);
		CGContextRelease(ctx);
		
		codi->w = (UInt32)w;
		codi->h = (UInt32)h;
		codi->len = w * h * 4;
		codi->data = (unsigned char*)pixels;
	} while(FALSE);
	SAFE_CF_RELEASE(url);
	SAFE_CF_RELEASE(imgSrc);
	SAFE_CF_RELEASE(img);
	return error;
}

int EncodeImageToFile(const char *uri, REBCDI *codi)
{
	int error = 0;
	CFStringRef type;
	CFMutableDictionaryRef prop;
	CGDataProviderRef data;
	CGColorSpaceRef colorSpace;
	CGImageRef img = NULL;
	CGImageDestinationRef imgDst = NULL;
	CFURLRef url;
	CFDataRef dataDst;
	
	switch (codi->type) {
	case CODI_IMG_PNG:  type = kUTTypePNG  ; break;     // Portable Network Graphics
	case CODI_IMG_JPEG: type = kUTTypeJPEG ; break;     // Joint Photographic Experts Group
	case CODI_IMG_JP2:  type = kUTTypeJPEG2000 ; break; // JPEG 2000
	case CODI_IMG_GIF:  type = kUTTypeGIF  ; break;     // Graphics Interchange Format
	case CODI_IMG_BMP:  type = kUTTypeBMP  ; break;     // Device independent bitmap
	case CODI_IMG_TIFF: type = kUTTypeTIFF ; break;     // Tagged Image File Format
	case CODI_IMG_HEIF: type = kSDUTTypeHEIC ; break;
	case CODI_IMG_DDS:  type = kSDUTTypeDDS  ; break;   // Microsoft DirectDraw Surface
	default:
			codi->error = 1;
			return codi->error;
	}
	do {
		data = CGDataProviderCreateWithData(NULL, codi->bits, codi->w * codi->h * 4, NULL);
		ASSERT_NOT_NULL(data, 1, "prepare input data");
		colorSpace = CGColorSpaceCreateDeviceRGB();
		img = CGImageCreate(codi->w, codi->h, 8, 32, codi->w * 4, colorSpace, (kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big), data, NULL, TRUE, 0);
		CGDataProviderRelease(data);
		CGColorSpaceRelease(colorSpace);
		ASSERT_NOT_NULL(img, 2, "create an image");
		
		if(uri == NULL) {
			// writing into preallocated buffer (fixed size!)
			dataDst = CFDataCreateWithBytesNoCopy(NULL, codi->data, codi->len, NULL);
			ASSERT_NOT_NULL(dataDst, 3, "prepare output data");
			imgDst = CGImageDestinationCreateWithData((CFMutableDataRef)dataDst, type, 1, 0);
		} else {
			// writing directly into file
			url = urlFromCString(uri);
			imgDst = CGImageDestinationCreateWithURL(url, type, 1, 0);
		}
		ASSERT_NOT_NULL(imgDst, 4, "create a destination image");
		// TODO: handle user defined options

		CGFloat quality = 0.6;
		CGImageDestinationAddImage(imgDst, img, (__bridge CFDictionaryRef)@{
			(__bridge NSString *)kCGImageDestinationLossyCompressionQuality: @(quality)
		});
		
		if(!CGImageDestinationFinalize(imgDst)) {
			// failed to finalize!
			error = 6;
		}
	} while(FALSE);
	if(uri == NULL && !error) {
		codi->len = CFDataGetLength(dataDst);
		if(codi->len == 0) error = 5;
		//CFRelease(dataDst);
	}
	SAFE_CF_RELEASE(img);
	SAFE_CF_RELEASE(imgDst);
	codi->error = error;
	return error;
}

File: src/os/win32/sys-codecs.cpp
/***********************************************************************
**
**  REBOL [R3] Language Interpreter and Run-time Environment
**
**	Copyright 2019-2023 Oldes
**
**  Licensed under the Apache License, Version 2.0 (the "License");
**  you may not use this file except in compliance with the License.
**  You may obtain a copy of the License at
**
**  http://www.apache.org/licenses/LICENSE-2.0
**
**  Unless required by applicable law or agreed to in writing, software
**  distributed under the License is distributed on an "AS IS" BASIS,
**  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
**  See the License for the specific language governing permissions and
**  limitations under the License.
**
************************************************************************
**
**  Title: Windows native codecs functionality
**  Author: Oldes
**  Purpose: To load and save images using system API
**	Related: core/n-image.c
**  Note: So far saves just one frame. For multi-image formats it would be
**        better to write a device (image:// port or something like that)
**
************************************************************************
**  Useful links:
**  https://chromium.googlesource.com/webm/webp-wic-codec
***********************************************************************/

#include "sys-codecs.h"

IWICImagingFactory *pIWICFactory = NULL;
static LARGE_INTEGER zero = { 0 };

CODECS_API int codecs_init()
{
	HRESULT hr = S_OK;
	if (!pIWICFactory)
	{
		//Create Windows Imaging Component ImagingFactory.
		hr = CoCreateInstance(
			CLSID_WICImagingFactory
			, NULL
			, CLSCTX_INPROC_SERVER
			, IID_PPV_ARGS(&pIWICFactory)
		);
	}
	return hr;
}

CODECS_API void codecs_fini()
{
	if (pIWICFactory) pIWICFactory->Release();
}


CODECS_API int DecodeImageFromFile(PCWSTR *uri, UINT frame, REBCDI *codi)
{
	HRESULT hr = S_OK;
	UINT w, h;
	BYTE *data = NULL;
	WICRect wrect;

	IWICBitmapDecoder     *pDecoder   = NULL;
    IWICBitmapFrameDecode *pSource    = NULL;
	IWICFormatConverter   *pConverter = NULL;
	IStream               *pStream    = NULL;

	do {
		hr = codecs_init();
		ASSERT_HR("codecs_init");

		if (uri) {
			TRACE("Filename %ls", uri);

			hr = pIWICFactory->CreateDecoderFromFilename(
				(LPCWSTR)uri
				, NULL
				, GENERIC_READ
				, WICDecodeMetadataCacheOnLoad
				, &pDecoder
			);
			ASSERT_HR("CreateDecoderFromFilename");
		}
		else {
			TRACE("Decoding from binary... bytes: %u", codi->len);

			// Global memory for stream will be released with the stream automaticaly
			HGLOBAL	hMem = ::GlobalAlloc(GMEM_MOVEABLE, codi->len);
			if (!hMem) {
				TRACE_ERR("GlobalAlloc");
				hr = GetLastError();
				break;
			}

			hr = CreateStreamOnHGlobal(hMem, TRUE, &pStream);
			ASSERT_HR("CreateStreamOnHGlobal");

			ULONG written;
			hr = pStream->Write(codi->data, codi->len, &written);
			ASSERT_HR("pStream->Write");

			// WIC JPEG decoder needs the stream to seek to head manually!
			// https://stackoverflow.com/a/12928336/494472
			pStream->Seek(zero, STREAM_SEEK_SET, NULL);

			hr = pIWICFactory->CreateDecoderFromStream(
				pStream
				, NULL
				, WICDecodeMetadataCacheOnLoad
				, &pDecoder
			);
			ASSERT_HR("CreateDecoderFromStream");
		}

		// Create the initial frame.
		UINT pCount = 0;
		pDecoder->GetFrameCount(&pCount);
		TRACE("Frames: %u", pCount);

		hr = pDecoder->GetFrame(frame, &pSource);
		ASSERT_HR("GetFrame");

		// Convert the image format to 32bppPBGRA
		hr = pIWICFactory->CreateFormatConverter(&pConverter);
		ASSERT_HR("CreateFormatConverter");

		hr = pConverter->Initialize(
			pSource
			, GUID_WICPixelFormat32bppBGRA
			, WICBitmapDitherTypeNone
			, NULL
			, 0.f
			, WICBitmapPaletteTypeMedianCut
		);
		ASSERT_HR("WIC Initialize pConverter");

		pConverter->GetSize(&w, &h);
		TRACE("size: %ux%u", w, h);
		codi->w = w;
		codi->h = h;
		codi->len = w * h * 4;

		data = (unsigned char *)malloc(codi->len);
		if (data) {
			wrect = { 0,0,(INT)w,(INT)h };
			hr = pConverter->CopyPixels(&wrect, w * 4, codi->len, data);
			ASSERT_HR("CopyPixels");
		}
	} while(FALSE);

	codi->error = hr;
	if (FAILED(hr)) {
		if (data) free(data); // in case that CopyPixels fails
	}
	else {
		codi->data = data;
	}

	RELEASE(pStream);
	RELEASE(pDecoder);
    RELEASE(pSource);
	RELEASE(pConverter);

	return hr;
}

HRESULT AddBoolProperty(IPropertyBag2 *pPropertybag, LPOLESTR name, VARIANT_BOOL value) {
	PROPBAG2 option = { 0 };
	option.pstrName = name;
	VARIANT varValue;    
	VariantInit(&varValue);
	varValue.vt = VT_BOOL;
	varValue.boolVal = value;      
	return pPropertybag->Write(1, &option, &varValue);   
}

CODECS_API int EncodeImageToFile(PCWSTR *uri, REBCDI *codi)
{
	HRESULT hr = S_OK;
//	UINT  w, h;
	UINT  size;
	BYTE *data = NULL;
	WICRect wrect;

//	ULONG bytes;

	IWICBitmap         *pWICBitmap     = NULL;
	IWICBitmapLock     *pWICBitmapLock = NULL;

	IWICBitmapEncoder     *encoder     = NULL;
	IWICStream            *wicStream   = NULL;
	IWICBitmapFrameEncode *frameEncode = NULL;
	IStream               *outStream   = NULL;

	IPropertyBag2 *pPropertybag = NULL;

	GUID containerFormat;

	HGLOBAL	hMem = NULL;

	switch (codi->type) {
	case CODI_IMG_PNG:  containerFormat = GUID_ContainerFormatPng  ; break;     // Portable Network Graphics
	case CODI_IMG_JPEG: containerFormat = GUID_ContainerFormatJpeg ; break;     // Joint Photographic Experts Group
	case CODI_IMG_GIF:  containerFormat = GUID_ContainerFormatGif  ; break;     // Graphics Interchange Format
	case CODI_IMG_DDS:  containerFormat = GUID_ContainerFormatDds  ; break;     // DirectDraw Surface
	case CODI_IMG_DNG:  containerFormat = GUID_ContainerFormatAdng  ; break;    // Digital Negative
	case CODI_IMG_BMP:  containerFormat = GUID_ContainerFormatBmp  ; break;     // Device independent bitmap
	case CODI_IMG_TIFF: containerFormat = GUID_ContainerFormatTiff ; break;     // Tagged Image File Format
	case CODI_IMG_JXR:  containerFormat = GUID_ContainerFormatWmp  ; break;     // Windows Digital Photo (JpegXR)
	case CODI_IMG_HEIF: containerFormat = GUID_ContainerFormatHeif ; break;     // High Efficiency Image Format
	case CODI_IMG_WEBP: containerFormat = GUID_ContainerFormatWebp ; break;     //
	default: return WINCODEC_ERR_COMPONENTNOTFOUND;
	}

	do {
		hr = codecs_init();
		ASSERT_HR("codecs_init");

		hr = pIWICFactory->CreateBitmap(
			  codi->w
			, codi->h
			, GUID_WICPixelFormat32bppBGRA
			, WICBitmapCacheOnLoad
			, &pWICBitmap
		);
		ASSERT_HR("CreateBitmap");

		wrect = {0,0,(INT)codi->w,(INT)codi->h};

		hr = pWICBitmap->Lock(&wrect, WICBitmapLockWrite, &pWICBitmapLock);
		ASSERT_HR("WICBitmap->Lock");

		hr = pWICBitmapLock->GetDataPointer(&size, &data);
		ASSERT_HR("GetDataPointer");

		hr = memcpy_s(data, size, codi->bits, size);
		ASSERT_HR("memcpy_s to pWICBitmap");
		RELEASE(pWICBitmapLock);
		
		hr = pIWICFactory->CreateEncoder(containerFormat, nullptr, &encoder);
		ASSERT_HR("pIWICFactory->CreateEncoder");

		if (uri == NULL) {
			// output to raw binary
			hr = CreateStreamOnHGlobal(NULL, TRUE, &outStream);
			ASSERT_HR("CreateStreamOnHGlobal");
		
		} else {
			// Create a stream for the encoder
			hr = pIWICFactory->CreateStream(&wicStream);
			ASSERT_HR("pIWICFactory->CreateStream");
			// Initialize the stream using the output file path
			hr = wicStream->InitializeFromFilename((LPCWSTR)uri, GENERIC_WRITE);
			ASSERT_HR("wicStream->InitializeFromFilename");
			hr = wicStream->QueryInterface(IID_PPV_ARGS(&outStream));
			ASSERT_HR("wicStream->QueryInterface");
		}
		// Create encoder to write to image file
		encoder->Initialize(outStream, WICBitmapEncoderNoCache);

		hr = encoder->CreateNewFrame(&frameEncode, &pPropertybag);
		ASSERT_HR("encoder->CreateNewFrame");

		//TODO: handle optional per-codec properties here!

		if (codi->type == CODI_IMG_BMP) {
			// force BMP encoder to always write alpha channel so far
			hr = AddBoolProperty(pPropertybag, (LPOLESTR)L"EnableV5Header32bppBGRA", VARIANT_TRUE);
			ASSERT_HR("Set EnableV5Header32bppBGRA");
		}

		//Initialize the encoder
		hr = frameEncode->Initialize(pPropertybag);
		ASSERT_HR("frameEncode->Initialize");
		frameEncode->SetSize(codi->w, codi->h);

		// Copy updated bitmap to output
		hr = frameEncode->WriteSource(pWICBitmap, nullptr);
		ASSERT_HR("frameEncode->WriteSource");
		hr = frameEncode->Commit();
		if(FAILED(hr)) break;
		hr = encoder->Commit();
		if(FAILED(hr)) break;

		outStream->Commit(STGC_DEFAULT);
		RELEASE(encoder);

		if (uri == NULL) {
			// copy data from stream to codi
			STATSTG stat;
			hr = outStream->Stat(&stat, STATFLAG_NONAME);
			ASSERT_HR("outStream->Stat");
			if (stat.cbSize.HighPart > 0 || stat.cbSize.LowPart > UINT_MAX) {
				hr = E_OUTOFMEMORY;
				break;
			}
			TRACE("Result has %lu bytes", stat.cbSize.LowPart);

			if (codi->len < stat.cbSize.LowPart) {
				// result binary series is not large enough!

				// TODO: we could allocate a new result buffer now and let the C side to
				// know, that result binary must be extended and data copied and freed.
				// or result could keep the stream, let C know, that more extension is needed,
				// than go back here, read data and release the stream.
				hr = E_OUTOFMEMORY;
				break;
			}

			ULONG read;
			outStream->Seek(zero, STREAM_SEEK_SET, NULL);
			hr = outStream->Read(codi->data, stat.cbSize.LowPart, &read);
			ASSERT_HR("outStream->Read");
			codi->len = read;
		}
	} while(FALSE);

	RELEASE(pPropertybag);
	RELEASE(pWICBitmap);
	RELEASE(pWICBitmapLock);
	RELEASE(outStream);
	RELEASE(frameEncode);
	RELEASE(wicStream);
	RELEASE(encoder);

	codi->error = hr;

	return hr;
}
